### PR TITLE
Manila Shares

### PR DESCRIPTION
**Pull request by Carthaca**
Initial manila implementation.

I plan to add snapshots and quota soon.

### CODE REVIEW

**Comment by Carthaca** [comment]
I disabled the `FrozenStringLiteralComment` cop - I don't think it is planned to go for ruby 3.0 in the near future

**Comment by gildub** [comment]
@Carthaca, LGTM.
We can merge once you have added the doco and examples.
Thanks

**Comment by Carthaca** [comment]
added docu + example and squashed the commits :)

**Comment by gildub** [comment]
The intent to add the resource for the documention is noble but I don't think that should belong to this patch. Could you please keep here only the documentation related to the Manila shares?

**Comment by Carthaca** [comment]
done - I will open a new PR with the documentation consolidation

File: docs/getting_started.md [hunk 1]
```diff
@@ -68,14 +68,15 @@ Fog::OpenStack.services
 These services can be explored in further depth in the following documents:
 
 * [Compute](compute.md)
+* [Introspection](introspection.md)
+* [Metering](metering.md)
+* [Network](network.md)
+* [NFV](nfv.md)
+* [Orchestration](orchestration.md)
+* [Planning](planning.md)
+* [Share (Manila)](share.md)
```
Comment (gildub) [inline docs/getting_started.md]: Documentation is effectively behind, thank you for doing this.
Meanwhile could you please make those 'common' changes into a separate Pull Request?
So we can keep track of changes.

File: docs/getting_started.md [hunk 2]
```diff
@@ -68,14 +68,15 @@ Fog::OpenStack.services
...
 ## Additional Resources
-* [fog.io](http://fog.io)
-* [Fog rdoc](http://rubydoc.info/gems/fog)
-* [Fog Github repo](https://github.com/fog/fog)
-* [Release Notes](https://github.com/fog/fog/blob/master/changelog.txt)
-
-
-## Support and Feedback
-Your feedback is appreciated! If you have specific issues with the **fog** SDK, you should file an [issue via Github](https://github.com/fog/fog/issues).
+[resources and feedback](shared/resources.md)
```
Comment (gildub) [inline docs/getting_started.md]: Same as above, please submit a separate PR for those changes not related to your initial Manila commit.
Also, I think, `shared` is confusing I would opt for `common` but it's only my view of it. :)

File: docs/share.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Share (Manila)
```
Comment (gildub) [inline docs/share.md]: I believe the generic component name is "Shared Filed System" not just "Share". 
Please change the name accordingly everywhere it needs to in order to reflect the full name, including in the file: `shared_file_system.md`. Thanks

File: docs/shared/connection_params.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+### Optional Connection Parameters
```
Comment (gildub) [inline docs/shared/connection_params.md]: As mentioned earlier, this one to go within a separate PR.
Also not keen on `shared` name.

File: examples/share/basics.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
+require 'fog/openstack'
```
Comment (gildub) [inline examples/share/basics.rb]: As above, rename examples/shared_file_system/basics.rb

File: lib/fog/network/openstack/requests/create_network.rb [hunk 6]
```diff
@@ -69,13 +69,13 @@ def create_network(options = {})
...
-            'shared'                => options[:shared],
+            'shared'                => options[:shared] || false,
             'subnets'               => [],
             'status'                => 'ACTIVE',
-            'admin_state_up'        => options[:admin_state_up],
+            'admin_state_up'        => options[:admin_state_up] || false,
             'tenant_id'             => options[:tenant_id],
             'qos_policy_id'         => options[:qos_policy_id],
-            'port_security_enabled' => options[:port_security_enabled]
+            'port_security_enabled' => options[:port_security_enabled] || false
```
Comment (gildub) [inline lib/fog/network/openstack/requests/create_network.rb]: Same as above. That's better but not in this PR.

File: lib/fog/openstack.rb [hunk 7]
```diff
@@ -106,6 +110,7 @@ module OpenStack
     service(:monitoring,    'Monitoring')
     service(:workflow,      'Workflow')
     service(:dns,           'DNS')
+    service(:share,         'Share')
```
Comment (gildub) [inline lib/fog/openstack.rb]: I actually initially missed that, I think the name could be `:shared_file_system`.

File: lib/fog/share/openstack.rb [hunk 8]
```diff
@@ -0,0 +1,239 @@
+module Fog
+  module Share
```
Comment (gildub) [inline lib/fog/share/openstack.rb]: To match a full generic name I think the module name `Share` would be something like SharedFileSystem. What do you think?

File: lib/fog/share/openstack/models/network.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
+require 'fog/openstack/models/model'
+
+module Fog
+  module Share
```
Comment (gildub) [inline lib/fog/share/openstack/models/network.rb]: Same as above, ShareFileSystem or so

**Comment by Carthaca** [comment]
Thanks for having a look.

To the network changes: I simply needed them to get the tests green :) You are right, that this should be kept separate, same with the common documentation changes. I will do that.

To the name: I used the service catalog as orientation, the service type of manila is 'share'. I like it better, because it is shorter than 'shared_filed_system'.

**Comment by Carthaca** [comment]
I am done with moving out the not manila related stuff.

Do we need a third opinion for the name or can we agree on 'share'?

**Comment by Carthaca** [comment]
Okay, it's now or never ^^

"Shared File System" is more specific.

**Comment by gildub** [comment]
Thank you for this great work!
